mlflow/models/evaluation/default_evaluator.py

```diff
@@ -494,28 +531,50 @@ def _log_model_explainability(self):
             f: f2 for f, f2 in zip(self.feature_names, truncated_feature_names)
         }
 
-        sampled_X = shap.sample(self.X, sample_rows)
+        sampled_X = shap.sample(self.X, sample_rows, random_state=0)
+        if algorithm == "kernel":
+            background_X = shap.sample(self.X, sample_rows, random_state=3)
+        else:
+            background_X = None
 
         if isinstance(sampled_X, pd.DataFrame):
             # For some shap explainer, the plot will use the DataFrame column names instead of
             # using feature_names argument value. So rename the dataframe column names.
             sampled_X = sampled_X.rename(columns=truncated_feature_name_map, copy=False)
+            if background_X is not None:
+                background_X = background_X.rename(columns=truncated_feature_name_map, copy=False)
+        else:
+            # If sample_X is numpy array, convert to pandas dataframe.
+            sampled_X = pd.DataFrame(sampled_X, columns=truncated_feature_name_map)
+            if background_X is not None:
+                background_X = pd.DataFrame(background_X, columns=truncated_feature_name_map)
 
         if algorithm:
-            supported_algos = ["exact", "permutation", "partition"]
-            if algorithm not in supported_algos:
-                raise ValueError(
-                    f"Specified explainer algorithm {algorithm} is unsupported. Currently only "
-                    f"support {','.join(supported_algos)} algorithms."
+            if algorithm == "kernel":
+                kernel_link = self.evaluator_config.get("explainability_kernel_link", "identity")
+                if kernel_link not in ["identity", "logit"]:
+                    raise ValueError(
+                        f"explainability_kernel_link config can only be set to 'identity' or 'logit'."
+                    )
+                mode_or_mean_dict = _compute_df_mode_or_mean(self.X)
+                mode_or_mean_dict = {truncated_feature_name_map[k]: v for k, v in mode_or_mean_dict}
+                sampled_X = sampled_X.fillna(mode_or_mean_dict)
+                background_X = background_X.fillna(mode_or_mean_dict)
+                predict_fn = lambda x: self.predict_fn(pd.DataFrame(x, columns=sampled_X.columns))
+                explainer = shap.KernelExplainer(predict_fn, background_X, link=kernel_link)
```

Comment:
```suggestion
                explainer = shap.KernelExplainer(self.predict_fn, background_X, link=kernel_link)
```

What would happen if we directly pass `self.predict_fn`?

Reply:
If so,
On the `mlflow pipeline evaluate` example test data, one error will be raised:
```
  File "/Users/weichen.xu/work/mlflow-private/mlflow/models/evaluation/default_evaluator.py", line 555, in _log_model_explainability
    explainer = shap.KernelExplainer(self.predict_fn, background_X, link=kernel_link)
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/shap/explainers/_kernel.py", line 69, in __init__
    model_null = match_model_to_data(self.model, self.data)
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/shap/utils/_legacy.py", line 112, in match_model_to_data
    out_val = model.f(data.data)
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/sklearn/utils/metaestimators.py", line 113, in <lambda>
    out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)  # noqa
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/sklearn/pipeline.py", line 469, in predict
    Xt = transform.transform(Xt)
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py", line 748, in transform
    Xs = self._fit_transform(
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py", line 606, in _fit_transform
    return Parallel(n_jobs=self.n_jobs)(
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/joblib/parallel.py", line 1043, in __call__
    if self.dispatch_one_batch(iterator):
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/joblib/parallel.py", line 833, in dispatch_one_batch
    islice = list(itertools.islice(iterator, big_batch_size))
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/sklearn/compose/_column_transformer.py", line 609, in <genexpr>
    X=_safe_indexing(X, column, axis=1),
  File "/Users/weichen.xu/opt/miniconda3/envs/py39/lib/python3.9/site-packages/sklearn/utils/__init__.py", line 370, in _safe_indexing
    raise ValueError(
ValueError: Specifying the columns using strings is only supported for pandas DataFrames
```

The error is caused by: the `predict_fn` passed to "shap.KernelExpainer" was called with a "Numpy array" type argument.

Reply:
got it 👍